Create GC-07-13.md (#827)
WebAssembly/meetings: gc/2021/GC-07-13.md

- **Location**: *link on calendar invite*

### Registration

None required if you've attended before. Fill out the form here to sign up if
it's your first time: https://forms.gle/JehrAB4gWbtHjybt9. The meeting is open
to CG members only.

## Logistics

The meeting will be on a zoom.us video conference.
Installation is required, see the calendar invite.

## Agenda items

1. Opening, welcome and roll call
    1. Opening of the meeting
    1. Introduction of attendees
1. Find volunteers for note taking (acting chair to volunteer)
1. Adoption of the agenda
1. Proposals and discussions
    1. Discussion: Problems with nominal types [Andreas Rossberg] (1 hour)
1. Closure

## Meeting Notes

### Introduction of attendees

- Thomas Lively
- Tim Steenvoorden
- Manos Koukoutos
- Ross Tate
- Alon Zakai
- Conrad Watt
- Andreas Rossberg
- Francis McCabe
- Zalim Bashorov
- Ryan Hunt
- Emanuel Ziegler
- Adam Klein
- Daniel Wirtz
- Asumu Takikawa
- Slava Kuzmich
- Jakob Kummerow
- Keith Miller
- Luke Wagner
- Fil Pizlo
- Deepti Gandluri
- Petr Penzin
- Paulo Matos
- Ioanna Dimitriou

### Discussion: Problems with nominal types [Andreas Rossberg] (1 hour)

> TODO: slides

Author: @rossberg, can you upload and link to your slides?

Reviewer: Since they were already made, it would be useful to see the remainder of the slides ahead of next week's meeting. Knowing what's coming helps people know if a question or concern will be addressed in a later slide or whether it needs to be discussed at the moment while it is on topic, which makes for a more efficient meeting.

AR: The different degrees of modularity are part of the value proposition of Wasm in my view.

RT: These are components like Luke explained, right?

> AR: No, these components could be anywhere on the range of coupling.

Reviewer 2: ```suggestion
AR: Not necessarily, these components could be anywhere on the range of coupling from the previous slide.
```

CW: But we’re drawing a distinction between components and modules. If you get too far away from Luke’s model, that might not be useful any more?

> AR: Not sure what you mean. There are two things. The component model that does not commit to an abstraction layer and then also the interface types things which establishes a particular abstraction.

Reviewer 2: ```suggestion
AR: Not sure what you mean. There are two things Luke proposed. The module linking proposal that does not commit to an abstraction layer and then interface types which establishes a particular abstraction.
```

> RT: In the highly-coupled setting, one of these components might be a language runtime.

Reviewer 2: ```suggestion
RT: In the highly-coupled setting, one of these components might be a language runtime.

AR: Yes, absolutely.
```

AR: Fair point, but I would argue that dynamic linking can be simple. Things are complicated because we are emulating native ld, not by necessity.

TL: That complexity is necessary to support real-world ecosystems when not spinning up a new language/ecosystem with the freedom to do the simple thing.

RT: Are you saying it’s currently possible to do deferred linking without knowing types?

AR: Yes, you link the RTTs the same way you link the functions.

RT: In my example, to create a Foo, you need the Rtt for Foo. In many languages you don’t need the dependent Bar to be loaded first. High level point is that you’re overstating what is possible.

AR: The point is that this [deferred linking] is a non-goal because we don’t expect to support it without overhead.

TL: I hope we can all agree that these are non-goals for the MVP, regardless of how well we could support it in the future.

RT: The J2CL team requested deferred loading.

CW: It’s not clear that they meant the kind of deferred loading that AR is saying is a non-goal here.

> RT: I’m not comfortable declaring this a non-goal without more feedback from the community.

Reviewer 2: ```suggestion
RT: I’m not comfortable declaring this a non-goal without more feedback from the community.

AR: At least it is for the MVP.
```

CW: When you say imports are an implementation detail of each module, it seems that even with equirecursive imports, if you had a convention of not importing anything, then you still need to care about what the other module is importing because it may have a different convention that requires e.g. importing RTTs.

> AR: I’m talking about a module system where you’re not explicitly instantiating modules. The instantiation is an implementation detail.

Reviewer 2: ```suggestion
AR: I’m talking about a typical module system where you’re not explicitly instantiating modules. The instantiation is an implementation detail.
```

CW: So this is just describing the user experience at the source level?

> AR: But perhaps also at the module level.

Reviewer 2: ```suggestion
AR: Yes, a language module model, but also a lower-level import semantics.
```

TL: What about the linker/loader that has the global knowledge that may or not be more Wasm?

RT: There just needs to be a convention that allows the linker/loader to load everything.

CW: To cut to the chase, this is going to lead back to the discussion of whether nominal types allow for such a convention.

CW: I appreciate the distinction between the point of view of the app, registry, user, linker, etc. But the registry needs to know the full client set eventually, but it might grow over time.

AR: Yes, but critically it does not know the full client set up front.

RT: The loader I sketched with nominal + staged compilation looks similar and can support.

CW: Can also just keep track of the things exported so far?

TL: It seems that the big picture is that AR is getting to the point that equirecursive types are more expressive and able to easily handle these dynamic repl-type situations. It seems that RT’s response is that the staged type linking

RT: If two modules are independent, then they only share runtime types. If they aren’t dependent, then they can import / export from each other.

> AR: That’s false.

Reviewer 2: ```suggestion
AR: That conclusion is false.
```

> RT: It would be great to see an example.

Reviewer 2: ```suggestion
RT: It would be great to see an example.

AR: Getting to it.
```

CW: It’s ok if nominal typing is less expressive, as long as it’s less expressive in ways we don’t care about.

TL: To be clear, most source level languages have structural types that do not rise to the expressivity of full equirecursive types.

CW: It also seems that you don’t need binary compatibility of the structural types. For example in OCaml and Haskell.

> AR: But yes, you can separately compile OCaml and Haskell modules and have them agree.

Reviewer 2: ```suggestion
AR: No, of course you do, you can separately compile OCaml and Haskell modules and they need to agree.
```

RT: The reason is because there is a runtime type constructor.

> AR: At some point in the compilation the types are erased, but that doesn’t make the compatibility problem go away.

Reviewer 2: ```suggestion
AR: In native compilation, at some point in the pipeline types are erased, but that doesn’t make the compatibility problem go away. In Wasm we cannot erase types.
```

RT: In OCaml, the function type A - > B compiles down to OCamlRef.

> AR: Yes, but then you have to downcast in the function and the caller and callee still have to agree on the type structure.

Reviewer 2: ```suggestion
AR: Yes, but then you have to downcast where you use the function and the caller and callee still have to agree on the type structure.
```

RT: The runtime package can provide the package type.

AR: Not if they’re infinite in number, for example the infinite space of tuple types.

RT: The only example so far of infinite structural types is tuples. Many runtimes lower tuples to max sizes.

AR: But it’s still exponential in the width of the tuple because each element can be a different type.

RT: Without generics, you would just have “tuple of size 2” rather than “tuple of int and float”

> AR: I would definitely want “tuple of int and float”

Reviewer 2: ```suggestion
AR: I would definitely want to unbox a “tuple of int and float”
```

RT: Many systems work perfectly well with just “tuple of size 2.” Only C# does otherwise.

> AR: But wob uses “tuple of int and float”

Reviewer 2: ```suggestion
AR: The Wob compiler uses “tuple of int and float”. Waml unboxes all structure fields. This isn't uncommon. 
```
